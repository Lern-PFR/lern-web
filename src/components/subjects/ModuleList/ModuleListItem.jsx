import PropTypes from 'prop-types';
import styled from 'styled-components';
import { withTranslation } from 'react-i18next';

import routes from 'routes/keys';
import { moduleAccessLinkButton, moduleDescription, moduleListItem, moduleName } from 'theme/pages/subjects/subjectDetailsPage';
import { StyledListItem } from 'components/shared/styledElements';
import { BodyCopy, DoublePica } from 'components/shared/typography';
import { Link } from 'components/shared/navigation';
import { PrimaryLinkButton } from 'components/shared/buttons';

const StyledModuleListItem = styled(StyledListItem)(
	{
		'&:not(:first-child):before': {
			content: '""',
			border: 'solid 1px black',
			position: 'absolute',
			top: '-2.1em',
			height: '2em',
			left: '10em',
		},
	}
);

/**
 * @name ModuleListItem
 * @description An item used to display an overview of a module inside its subject's details page.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {string}	id			The current module's id.
 * @param {string}	name		The current module's name.
 * @param {string}	description	A short description of the current module.
 * @param {bool}	[disabled]	Set to true if the current module is not yet unlocked for the current user.
 * @param {func}	t			The translation method provided by the withTranslation HoC.
 */
const ModuleListItem = ({ id, name, description, disabled, t }) => (
	<StyledModuleListItem {...moduleListItem}>
		<DoublePica {...moduleName}>{name}</DoublePica>
		<BodyCopy {...moduleDescription}>{description}</BodyCopy>
		<PrimaryLinkButton {...moduleAccessLinkButton} disabled={disabled}>
			<Link to={routes.modules.moduleDetails.replace(':id', id)} disabled={disabled}>{t('subjects.links.module')}</Link>
		</PrimaryLinkButton>
	</StyledModuleListItem>
);

ModuleListItem.propTypes = {
	id: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	disabled: PropTypes.bool,
	t: PropTypes.func.isRequired,
};

ModuleListItem.defaultProps = {
	disabled: false,
};

export default withTranslation()(ModuleListItem);
