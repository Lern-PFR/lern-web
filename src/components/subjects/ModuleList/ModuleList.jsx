import PropTypes from 'prop-types';
import conf from 'conf';
import styled from 'styled-components';
import { StyledDiv, StyledList, StyledSvg } from 'components/shared/styledElements';
import { BodyCopy } from 'components/shared/typography';
import { moduleList, moduleListNoDataContainer, moduleListNoDataSvg, moduleListNoDataText } from 'theme/pages/subjects/subjectDetailsPage';
import ModuleListItem from './ModuleListItem';

const NoDataSvg = styled(StyledSvg)(
	{
		'& svg': moduleListNoDataSvg,
	},
);

/**
 * @name ModuleList
 * @description A component used to display a subject's module list.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {array}	modules		The list of modules to display.
 * @param {string}	noDataText	The message to display the modules array is empty.
 */
const ModuleList = ({ modules, noDataText }) => (
	<>
		{(!modules || modules.length === 0) && (
			<StyledDiv {...moduleListNoDataContainer}>
				<NoDataSvg src={`${conf.svgIllustrationsPath}/blank_canvas.svg`} />
				<BodyCopy {...moduleListNoDataText}>{noDataText}</BodyCopy>
			</StyledDiv>
		)}
		{modules && modules.length > 0 && (
			<StyledList {...moduleList} listStyle="none">
				{modules.map(({ id, ...moduleData }) => (
					<ModuleListItem
						key={id}
						id={id}
						{...moduleData}
					/>
				))}
			</StyledList>
		)}
	</>
);

ModuleList.propTypes = {
	modules: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.string.isRequired,
		title: PropTypes.string.isRequired,
		description: PropTypes.string.isRequired,
	})).isRequired,
	noDataText: PropTypes.string.isRequired,
};

export default ModuleList;
