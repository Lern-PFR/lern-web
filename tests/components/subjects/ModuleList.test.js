import { shallow } from 'enzyme';
import ModuleList from 'components/subjects/ModuleList';

describe('ModuleList', () => {
	it('should match previous snapshot', () => {
		const sut = (<ModuleList modules={[]} noDataText="" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
