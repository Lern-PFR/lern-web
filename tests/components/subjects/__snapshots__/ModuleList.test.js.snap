// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleList should match previous snapshot 1`] = `
<styled.ul
  display="grid"
  gridAutoRows="minmax(250px, 1fr)"
  gridRowGap="2em"
  listStyle="none"
/>
`;
